DEVELOPER_GUIDE.md

```diff
@@ -226,3 +226,70 @@ original PR with an appropriate label `backport <backport-branch-name>` is merge
+`src/main/java/org/opensearch/neuralsearch/rerank_factory/rerank_processor_factory.java`
+
+### Modular code
+
+Try to organize code into small classes and methods with a single concise purpose. It's preferable to have multiple small
+methods rather than a long single one and does everything.
+
+### Documentation
+
+Document you code. That includes purpose of new classes, every public method and code sections that have critical or non-trivial
+logic (check this example https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java#L238). 
+
+When you submit a feature PR, please submit a new
+[documentation issue](https://github.com/opensearch-project/documentation-website/issues/new/choose). This is a path for the documentation to be published as part of https://opensearch.org/docs/latest/ documentation site.
+
+Please be prepared to provide any additional guidance (like example of query request/response, details of API parameters etc.) for the team doing the documentation.
+
+### Code style
+
+For the most part, we're using common conventions for Java projects. Here are a few things to keep in mind.
+
+1. Use descriptive names for classes, methods, fields, and variables.
+2. Avoid abbreviations unless they are widely accepted
+3. Use `final` on all method arguments unless it's absolutely necessary 
+4. Wildcard imports are not allowed.
+5. Static imports are preferred over qualified imports when using static methods
+6. Prefer creating non-static methods whenever possible. Static methods should generally be avoided as they are often used as a shortcut.
+7. Use functional programming style inside methods unless it's a performance critical section. 
+8. For parameters of lambda expression please use meaningful names instead of shorten cryptic ones.
+9. Use Optional for return values if the value may not be present. This should be preferred to returning null.
+10. Do not create checked exceptions, and do not throw checked exceptions from public methods whenever possible. In general, if you call a method with a checked exception, you should wrap that exception into an unchecked exception.
+11. Throwing checked exceptions from private methods is acceptable.
+12. Use String.format style in case your string is using parameters, prefer that to direct string concatenation
+13. Prefer Lombok annotations to the manually written boilerplate code
+
+Some of mentioned styles are enforced by the `spotless` Gradle plugin, please check [Java Language Formatting Guidelines](##Java Language Formatting Guidelines) section for more details.
+
+### Tests
+
+Write unit and integration tests for your new functionality. 
+
+Unit tests are preferred as they are cheap and fast, try to use them to cover all possible
+combinations of parameters. Utilize mocks to mimic dependencies. 
+
+Integration tests should be used frugally only for main happy path scenario, or when it takes a lot of efforts to mock 
```

Comment:
For unhappy path we should have IT too where we should validate if correct response codes are sent to the customer. We should at-least add some negative cases if we are adding new APIs.

Reply:
ack, added mentioned points regarding the unhappy path